exporters/common/src/main/java/io/opentelemetry/exporter/internal/marshal/StatelessMarshalerUtil.java

```diff
@@ -299,16 +299,56 @@ public static <K, V> int sizeMessageWithContext(
   }
 
   /** Returns the size of utf8 encoded string in bytes. */
-  @SuppressWarnings("UnusedVariable")
   private static int getUtf8Size(String string, MarshalerContext context) {
-    return getUtf8Size(string);
+    return getUtf8Size(string, context.marshalStringUnsafe());
   }
 
   // Visible for testing
-  static int getUtf8Size(String string) {
+  static int getUtf8Size(String string, boolean useUnsafe) {
+    if (useUnsafe && UnsafeString.isAvailable() && UnsafeString.isLatin1(string)) {
+      byte[] bytes = UnsafeString.getBytes(string);
+      // latin1 bytes with negative value (most significant bit set) are encoded as 2 bytes in utf8
+      return string.length() + countNegative(bytes);
+    }
+
     return encodedUtf8Length(string);
   }
 
+  /** Returns the count of bytes with negative value. */
+  private static int countNegative(byte[] bytes) {
+    int count = 0;
+    int offset = 0;
+    // We are processing one long (8 bytes) at a time. In the inner loop we are keeping counts in a
+    // long where each byte in the long is a separate counter. Due to this the inner loop can
+    // process a maximum of 8*255 bytes at a time without overflow.
+    for (int i = 1; i <= bytes.length / (8 * 255) + 1; i++) {
```

Comment:
I think it could help readability to extract some of these constants and bit patterns to static helpers so the reader can key off the variable name to understand what's happening. It took me a bit to realize that 8*255 was significant because its the max size of each inter loop (despite the comment saying it right in front of me!), where 255 is the max size of each of the eight 8-bit counters contained inside the single 64-bit `long tmp`. Also that `0x8080808080808080L` was a bit mask for extracting the most significant bit of the 8 sequential bytes composing the long, and that `0xff` is a bitmask for extracting the 8 least significant bits representing each of the eight 8-bit counters contained inside the single 64-bit `long tmp`.